Dynamic imports don't work

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior

It should be possible to use dynamic imports, i.e. `import('my-module').then(module => ...)`

## Current Behavior

Rollup doesn't seem to generate correct output for import statements. It ends up removing the import function with this:

```
() => new Promise(function (resolve) { resolve(_interopNamespace(require('my-module'))); })
```

## Possible Solution

Changing the rollup output format to 'module' seems to produce correct code. We might want to switch to that and make sure it works with the rest of the build.

## Steps to Reproduce

Try to use a dynamic import in a plugin.

## Context

Dynamic imports is a huge part of keeping initial page load fast, it's best to have it working early on so that all plugins can use it.

This also requires an update to `tsconfig.json`.
```json
  "compilerOptions": {
    "module": "esnext"
  }
```
Build error without updating tsconfig, for a new plugin which has dynamic import:
```
src/index.ts:19:1 - error TS1323: Dynamic imports are only supported when the '--module' flag is set to 'es2020', 'esnext', 'commonjs', 'amd', 'system', or 'umd'.
```

--- a/packages/cli/config/jest.js
+++ b/packages/cli/config/jest.js
@@ -29,6 +29,13 @@ if (fs.existsSync('jest.config.js')) {
     moduleNameMapper: {
       '\\.(css|less|scss|sss|styl)$': require.resolve('jest-css-modules'),
     },
+    // We build .esm.js files with plugin:build, so to be able to load these in tests they need to be transformed
+    // TODO: jest is working on module support, it's possible that we can remove this in the future
+    transform: {
+      '\\.esm\\.js$': require.resolve('jest-esm-transformer'),
+    },
+    // Default behaviour is to not apply transforms for node_modules, but we still want to tranform .esm.js files
+    transformIgnorePatterns: ['/node_modules/(?!.*\\.esm\\.js$)'],
   };
 
   // Use src/setupTests.ts as the default location for configuring test env

--- a/packages/cli/config/tsconfig.json
+++ b/packages/cli/config/tsconfig.json
@@ -6,7 +6,8 @@
     "noEmit": false,
     "declarationMap": true,
     "incremental": true,
-    "module": "es6",
+    "target": "ES2019",
+    "module": "ESNext",
     "resolveJsonModule": true,
     "esModuleInterop": true,
     "types": ["node", "jest"]

--- a/packages/cli/package.json
+++ b/packages/cli/package.json
@@ -67,6 +67,7 @@
     "inquirer": "^7.0.4",
     "jest": "^25.1.0",
     "jest-css-modules": "^2.1.0",
+    "jest-esm-transformer": "^1.0.0",
     "ora": "^4.0.3",
     "react": "^16.0.0",
     "react-dev-utils": "^10.2.0",

--- a/packages/cli/src/commands/plugin/rollup.config.ts
+++ b/packages/cli/src/commands/plugin/rollup.config.ts
@@ -27,8 +27,8 @@ import { paths } from 'lib/paths';
 export default {
   input: 'src/index.ts',
   output: {
-    file: 'dist/index.cjs.js',
-    format: 'cjs',
+    file: 'dist/index.esm.js',
+    format: 'module',
   },
   plugins: [
     peerDepsExternal({

--- a/packages/cli/templates/default-app/plugins/welcome/package.json.hbs
+++ b/packages/cli/templates/default-app/plugins/welcome/package.json.hbs
@@ -1,7 +1,7 @@
 {
   "name": "plugin-welcome",
   "version": "0.0.0",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "private": true,
   "scripts": {

--- a/packages/cli/templates/default-plugin/package.json.hbs
+++ b/packages/cli/templates/default-plugin/package.json.hbs
@@ -1,7 +1,7 @@
 {
   "name": "@backstage/plugin-{{id}}",
   "version": "{{version}}",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "license": "Apache-2.0",
   "private": true,

--- a/packages/core/package.json
+++ b/packages/core/package.json
@@ -16,7 +16,7 @@
     "backstage"
   ],
   "license": "Apache-2.0",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "scripts": {
     "build": "backstage-cli plugin:build",

--- a/packages/dev-utils/package.json
+++ b/packages/dev-utils/package.json
@@ -16,7 +16,7 @@
     "backstage"
   ],
   "license": "Apache-2.0",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "scripts": {
     "build": "backstage-cli plugin:build",

--- a/packages/test-utils-core/package.json
+++ b/packages/test-utils-core/package.json
@@ -16,7 +16,7 @@
     "backstage"
   ],
   "license": "Apache-2.0",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "scripts": {
     "build": "backstage-cli plugin:build",

--- a/packages/test-utils/package.json
+++ b/packages/test-utils/package.json
@@ -16,7 +16,7 @@
     "backstage"
   ],
   "license": "Apache-2.0",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "scripts": {
     "build": "backstage-cli plugin:build",

--- a/packages/theme/package.json
+++ b/packages/theme/package.json
@@ -16,7 +16,7 @@
     "backstage"
   ],
   "license": "Apache-2.0",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "scripts": {
     "build": "backstage-cli plugin:build",

--- a/plugins/graphiql/package.json
+++ b/plugins/graphiql/package.json
@@ -16,7 +16,7 @@
     "backstage"
   ],
   "license": "Apache-2.0",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "scripts": {
     "build": "backstage-cli plugin:build",

--- a/plugins/graphiql/src/components/GraphiQLBrowser/GraphiQLBrowser.tsx
+++ b/plugins/graphiql/src/components/GraphiQLBrowser/GraphiQLBrowser.tsx
@@ -14,13 +14,15 @@
  * limitations under the License.
  */
 
-import React, { FC, useState } from 'react';
+import React, { FC, useState, Suspense } from 'react';
 import { Tabs, Tab, makeStyles, Typography, Divider } from '@material-ui/core';
 import 'graphiql/graphiql.css';
-import GraphiQL from 'graphiql';
 import { StorageBucket } from 'lib/storage';
 import { GraphQLEndpoint } from 'lib/api';
+import { Progress } from '@backstage/core';
 import { BackstageTheme } from '@backstage/theme';
+
+const GraphiQL = React.lazy(() => import('graphiql'));
 
 const useStyles = makeStyles<BackstageTheme>(theme => ({
   root: {
@@ -58,20 +60,22 @@ export const GraphiQLBrowser: FC<GraphiQLBrowserProps> = ({ endpoints }) => {
 
   return (
     <div className={classes.root}>
-      <Tabs
-        classes={{ root: classes.tabs }}
-        value={tabIndex}
-        onChange={(_, value) => setTabIndex(value)}
-        indicatorColor="primary"
-      >
-        {endpoints.map(({ title }, index) => (
-          <Tab key={index} label={title} value={index} />
-        ))}
-      </Tabs>
-      <Divider />
-      <div className={classes.graphiQlWrapper}>
-        <GraphiQL key={tabIndex} fetcher={fetcher} storage={storage} />
-      </div>
+      <Suspense fallback={<Progress />}>
+        <Tabs
+          classes={{ root: classes.tabs }}
+          value={tabIndex}
+          onChange={(_, value) => setTabIndex(value)}
+          indicatorColor="primary"
+        >
+          {endpoints.map(({ title }, index) => (
+            <Tab key={index} label={title} value={index} />
+          ))}
+        </Tabs>
+        <Divider />
+        <div className={classes.graphiQlWrapper}>
+          <GraphiQL key={tabIndex} fetcher={fetcher} storage={storage} />
+        </div>
+      </Suspense>
     </div>
   );
 };

--- a/plugins/home-page/package.json
+++ b/plugins/home-page/package.json
@@ -1,7 +1,7 @@
 {
   "name": "@backstage/plugin-home-page",
   "version": "0.1.1-alpha.4",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "license": "Apache-2.0",
   "private": true,

--- a/plugins/inventory/package.json
+++ b/plugins/inventory/package.json
@@ -1,7 +1,7 @@
 {
   "name": "@backstage/plugin-inventory",
   "version": "0.1.1-alpha.4",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "license": "Apache-2.0",
   "private": true,

--- a/plugins/lighthouse/package.json
+++ b/plugins/lighthouse/package.json
@@ -1,7 +1,7 @@
 {
   "name": "@backstage/plugin-lighthouse",
   "version": "0.1.1-alpha.4",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "license": "Apache-2.0",
   "private": true,

--- a/plugins/tech-radar/package.json
+++ b/plugins/tech-radar/package.json
@@ -1,7 +1,7 @@
 {
   "name": "@backstage/plugin-tech-radar",
   "version": "0.1.1-alpha.4",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "license": "Apache-2.0",
   "private": true,

--- a/plugins/welcome/package.json
+++ b/plugins/welcome/package.json
@@ -1,7 +1,7 @@
 {
   "name": "@backstage/plugin-welcome",
   "version": "0.1.1-alpha.4",
-  "main": "dist/index.cjs.js",
+  "main": "dist/index.esm.js",
   "types": "dist/index.d.ts",
   "private": true,
   "license": "Apache-2.0",

